yazi-core/src/tab/selected.rs

```diff
@@ -0,0 +1,223 @@
+use std::collections::{BTreeSet, HashMap};
+
+use yazi_shared::fs::Url;
+
+#[derive(Default)]
+pub struct Selected {
+	inner:   BTreeSet<Url>,
+	parents: HashMap<Url, usize>,
+}
+
+impl Selected {
+	pub fn get_inner(&self) -> BTreeSet<Url> { self.inner.clone() }
+
+	pub fn insert(&mut self, url: Url) -> bool { self.insert_many(&[&url]) }
+
+	pub fn insert_many(&mut self, urls: &[&Url]) -> bool {
+		if urls.is_empty() {
+			return true;
+		}
+
+		let mut parent = urls[0].parent_url();
+		while let Some(u) = parent {
+			if self.inner.contains(&u) {
+				return false;
+			}
+			parent = u.parent_url();
+		}
+
+		if self.parents.contains_key(urls[0]) {
+			return false;
+		}
+
+		let mut parent = urls[0].parent_url();
+		while let Some(u) = parent {
+			parent = u.parent_url();
+			*self.parents.entry(u).or_insert(0) += urls.len();
+		}
+
+		self.inner.extend(urls.iter().cloned().cloned());
+		true
+	}
+
+	pub fn remove(&mut self, url: &Url) -> bool {
+		if !self.inner.remove(url) {
+			return false;
+		}
+
+		let mut current_path = url.clone();
+		while let Some(parent) = current_path.parent_url() {
+			current_path = parent;
+
+			let counter = self.parents.entry(current_path.clone()).or_insert(0);
+			*counter -= 1;
```

Comment:
When `current_path` does not exist, inserting `0` and then `0 -= 1;` will cause an overflow.